### PR TITLE
[AOTInductor] Implement autograd eager backend for native triton kernels

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110486
* __->__ #110403

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -194,6 +194,18 @@ def wrap_with_proxy(e, proxy, constant):
             # example use case: allreduce_ returns ([tensor], work)
             for idx, ee in enumerate(e):
                 wrap_with_proxy(ee, proxy[idx], get_constant(idx))
+        elif isinstance(e, dict):
+            assert constant is None
+
+            if isinstance(proxy, fx.Proxy):
+                set_meta(proxy, e)
+
+            # example use case: allreduce_ returns ([tensor], work)
```
Comment (oulgen) [inline torch/fx/experimental/proxy_tensor.py:203]: i will update this comment

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 2]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+def trace_triton_kernel_wrapper(proxy_mode, func_overload, *, kernel, grid, kwargs):
+    with disable_proxy_modes_tracing():
+        out = func_overload(kernel=kernel, grid=grid, kwargs=kwargs)
+
+    fn = functools.partial(func_overload, kernel=kernel)
+    # FX graph needs __name__ and __module__ attributes
+    fn.__name__ = func_overload.__name__  # type:ignore[attr-defined]
+    if not hasattr(fn, "__module__"):
+        # Super hacky but on AMD __module__ is not set
+        fn.__module__ = "itertools"
```
Comment (zou3519) [inline torch/_higher_order_ops/triton_kernel_wrap.py:54]: Could we unify this logic with the Dynamo logic that does something similar? Maybe put into a helper function somewhere? Also we might want to note somewhere that we are doing a hack to workaround how fx does not allow functions in the graph.

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 3]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+
+    node_args = {"grid": grid, "kwargs": kwargs}
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function",
+        fn,
+        (),
+        proxy_args,
+        name="triton_kernel_wrapper_mutation",
```
Comment (zou3519) [inline torch/_higher_order_ops/triton_kernel_wrap.py:63]: nit: the name should probably be `overload.__name__`; in the functional case it is better for the name of the output to be be `triton_kernel_wrapper_functional` than `triton_kernel_wrapper_mutation`

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 4]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+        return ctx.wrap_tensors(outputs)
+
+
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.PythonDispatcher)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.PythonTLSSnapshot)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.ADInplaceOrView)
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.BackendSelect)
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutocastCPU)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutocastCUDA)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutogradCUDA)
```
Comment (zou3519) [inline torch/_higher_order_ops/triton_kernel_wrap.py:158]: What happens when we encounter a CPU Tensor (and try to pass it to this triton kernel underneath torch.compile?)

We might need to update this to `triton_kernel_wrapper_mutation.fallthrough(DispatchKey.Autograd)`, if that works. Otherwise we should probably also add a `triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutogradCPU)`.

File: torch/_subclasses/functional_tensor.py [hunk 5]
```diff
@@ -170,6 +170,15 @@ def from_functional(self):
         torch._sync(self)
         return torch._from_functional_tensor(self.elem)
 
+    def replace_(self, output) -> None:
+        torch._functionalize_replace(self.elem, output)  # type: ignore[attr-defined]
```
Comment (zou3519) [inline torch/_subclasses/functional_tensor.py:174]: Nit: add the type annotation to https://github.com/pytorch/pytorch/blob/main/torch/_C/__init__.pyi.in to make mypy happy, but it looks like we didn't care much about that in this file

**Review by zou3519 — APPROVED** [review]
LGTM, minor nits

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 6]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+    fn = functools.partial(func_overload, kernel=kernel)
+    # FX graph needs __name__ and __module__ attributes
+    fn.__name__ = func_overload.__name__  # type:ignore[attr-defined]
+    if not hasattr(fn, "__module__"):
+        # Super hacky but on AMD __module__ is not set
+        fn.__module__ = "itertools"
+
+    node_args = {"grid": grid, "kwargs": kwargs}
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
```
Comment (bdhirsh) [inline torch/_higher_order_ops/triton_kernel_wrap.py:58]: Comment that is probably more relevant for the previous dynamo PR (although fine to deal with as a followup): It's probably worth having a test for the case where the triton kernel (and maybe the grid lambda) is an inner function. Does everything still work out? (The graph that we compile probably directly holds a reference that inner function, so will the lifetime of the inner function match what its lifetime would have been in eager mode?)

```
@torch.compile
def f(x, y):
        @triton.jit
        def kernel(in_ptr, out_ptr0, BLOCK_SIZE: "tl.constexpr"):
            pid = tl.program_id(axis=0)
            block_start = pid * BLOCK_SIZE
            ...
        # if grid/kernel hold onto some global state (like a large cuda tensor), then that memory will stick around until grid/kernel is cleaned up.
        return kernel[grid](x)
    ....
```

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 7]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+
+@triton_kernel_wrapper_mutation.py_impl(FakeTensorMode)
+def triton_kernel_wrapper_mutation_fake_tensor_mode(mode, *, kernel, grid, kwargs):
+    with mode:
+        return None
+
+
+def trace_triton_kernel_wrapper(proxy_mode, func_overload, *, kernel, grid, kwargs):
+    with disable_proxy_modes_tracing():
+        out = func_overload(kernel=kernel, grid=grid, kwargs=kwargs)
```
Comment (bdhirsh) [inline torch/_higher_order_ops/triton_kernel_wrap.py:47]: I believe this is always supposed to return `None`, right? Since this will redispatch to the `FakeTensorMode` implementation, which returns None (since this higher order op is only ever used for triton kernels).

Maybe add an assert?

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 8]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+            grid=grid,
+            kwargs=kwargs,
+        )
+    else:
+        triton_kernel_wrapper_mutation(kernel=kernel, grid=grid, kwargs=kwargs)
+
+    return None
+
+
+@triton_kernel_wrapper_mutation.py_functionalize_impl
```
Comment (bdhirsh) [inline torch/_higher_order_ops/triton_kernel_wrap.py:86]: In this implementation, we're also relying on the fact that both `grid` and `kernel` will never mutate the metadata of any of the inputs. I think we said we'll (eventually) try to check for mutations from `grid` in dynamo and error. But is it possible for a triton kernel to mutate the metadata of a tensor input? e.g. `in_tensor.transpose_(1, 0)`.

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 9]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+        "call_function",
+        fn,
+        (),
+        proxy_args,
+        name="triton_kernel_wrapper_mutation",
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@triton_kernel_wrapper_mutation.py_impl(ProxyTorchDispatchMode)
```
Comment (bdhirsh) [inline torch/_higher_order_ops/triton_kernel_wrap.py:68]: To be fair, I don't think we'll ever actually hit this in the "vanilla" torch.compile flow: functionalization runs above `ProxyTorchDispatchMode`, so we will always have converted `triton_kernel_wrapper_mutation` calls into `triton_kernel_wrapper_functional` calls in the graph before hitting the proxy mode.

Mostly just an fyi - still agreed that we need this (for example, pre_dispatch tracing today will hit this, since it does not (yet) use functionalization)

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 10]
```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
...
+        ctx.replace(input_arg, output_arg)
+        ctx.commit_update(input_arg)
+        ctx.sync(input_arg)
+    return None
+
+
+@triton_kernel_wrapper_functional.py_impl(DispatchKey.CompositeExplicitAutograd)
+def triton_kernel_wrapper_functional_dense(*, kernel, grid, kwargs):
+    kwargs = {
+        key: (clone_preserve_strides(val) if isinstance(val, Tensor) else val)
```
Comment (bdhirsh) [inline torch/_higher_order_ops/triton_kernel_wrap.py:110]: One comment I made to Richard: maybe we should add a TODO here: eventually, for performance reasons we probably want to ensure that these `clone_preserve_strides` calls are **never** executed at runtime (inductor should always optimize them away).

We can't actually guarantee this today though, until we fix this issue: https://github.com/pytorch/pytorch/issues/109240

File: torch/fx/experimental/proxy_tensor.py [hunk 11]
```diff
@@ -194,6 +194,18 @@ def wrap_with_proxy(e, proxy, constant):
             # example use case: allreduce_ returns ([tensor], work)
             for idx, ee in enumerate(e):
                 wrap_with_proxy(ee, proxy[idx], get_constant(idx))
+        elif isinstance(e, dict):
+            assert constant is None
```
Comment (bdhirsh) [inline torch/fx/experimental/proxy_tensor.py:198]: Maybe worth a comment: in theory we could support const-prop when proxy-tensor-tracing operators that returns dicts of tensors, but we have no use case for it today (since the only op we currently trace that can return a dict is `triton_kernel_wrapper_functional/mutation`, which does not participate in const-prop)

File: test/dynamo/test_functions.py [hunk 12]
```diff
@@ -1541,6 +1541,73 @@ def add_kernel(
         # Make sure it is NOT modified
         self.assertEqual(output, torch.zeros_like(t1))
 
+    @requires_cuda()
+    @requires_triton()
+    def test_triton_kernel_functionalize(self):
```
Comment (bdhirsh) [inline test/dynamo/test_functions.py:1546]: I think these tests are fine for this PR, but when we add inductor support it would be great to have some more comprehensive tests (just testing that the outputs match eager mode):

(1) triton kernel that mutates an input to the graph
(2) triton kernel that mutate a view of an input to the graph
(3) graph that returns a view/alias of one of the mutated arguments to a triton kernel

For the first case (simple input mutation), there are ~3 individual cases worth testing. This is because these cases impact whether or not AOTAutograd sends the `copy_()` (input mutation) directly to inductor in the graph, so they affect whether or not inductor will be able to optimize away the clone.

(a) the graph is an inference graph (test is under `no_grad()`, or no inputs to the graph require gradient)
(b) graph is a training graph (at least one input requires grad), **but** the arguments to the triton kernel to not require grad
(c) graph is a training graph, and the argument(s) to the triton kernel also require grad.

**Review by bdhirsh — APPROVED** [review]
looks good! Left some minor comments that don't all need to be addressed in this PR

File: torch/_higher_order_ops/triton_kernel_wrap.py (hunk 7 above)
Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:47]: Only the mutation version returns none, but the functional version returns kwargs

File: test/dynamo/test_functions.py (hunk 12 above)
Comment (oulgen) [reply test/dynamo/test_functions.py:1546]: Will add as follow up. Although I need some pointers for how to implement some of these test cases. 

Right now all my tests are (1) I believe and none of my tensors require grad (i did not do require_grad=True).
* How do make inputs to triton kernel require grad?
* Do I need to test things differently when tensors require grad or do I still check outputs match?

File: torch/_higher_order_ops/triton_kernel_wrap.py (hunk 6 above)
Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:58]: Will also do as follow up

File: torch/_higher_order_ops/triton_kernel_wrap.py (hunk 8 above)
Comment (oulgen) [reply torch/_higher_order_ops/triton_kernel_wrap.py:86]: We do the clone/replace for the triton kernel inputs, if the kernel mutates the metadata will that not be sufficient? @jansel any comments on this?

As for the grid, @jansel recommended doing something similar to what cond does for mutation checking, so i was planning on doing something like that.

Comment (bdhirsh) [reply torch/_higher_order_ops/triton_kernel_wrap.py:92]: This FakeTensorMode rule for the functional op:
```
    with mode:
        return {
            key: (clone_preserve_strides(val) if isinstance(val, Tensor) else val)
            for key, val in kwargs.items()
        }
```

Is implicitly assuming that the outputs to the triton kernel always have the same metadata (sizes/strides/etc) of the inputs. So if we're handed a triton kernel that mutates sizes/strides of a tensor input, our shape prop logic will be wrong during tracing.

To be fair though - I'm imagining that it's either impossible or very rare for a triton kernel to do this - so if it is possible, the answer is probably to detect/ban it in dynamo.

File: test/dynamo/test_functions.py (hunk 12 above)
Comment (bdhirsh) [reply test/dynamo/test_functions.py:1546]: > How do make inputs to triton kernel require grad?

You can just set your tensor inputs to the compiled code to require grad:
```
@torch.compile
def f(x):
    return kernel[grid](x, ...)

x = torch.randn(4, requires_grad=True)
out = f(x)
```

> Do I need to test things differently when tensors require grad or do I still check outputs match?

Testing just the forward is fine - but you're right, a more comprehensive test would involve also running autograd and checking that the gradients match between eager and compile. Example:
```
@torch.compile
def f_triton_input_mutation_of_view(x, out):
    out_view = out[:, 1]
    kernel[grid](x=x, out=out_view)
    return torch.mul(out_view, 2)

x = torch.randn(4)
out = torch.randn(4, requires_grad=True)
# run autograd
out.sum().backward()

# Run the same with ref inputs on eager, check that the grads match
self.assertEqual(out.grad, out_ref.grad)
```
